src/detect-transform-base64.c

```diff
@@ -37,33 +37,48 @@
 #include "util-unittest.h"
 #include "util-print.h"
 
-static int DetectTransformFromBase64DecodeSetup(DetectEngineCtx *, Signature *, const char *);
-static void DetectTransformFromBase64DecodeFree(DetectEngineCtx *, void *);
 #ifdef UNITTESTS
 #define DETECT_TRANSFORM_FROM_BASE64_MODE_DEFAULT (uint8_t) SCBase64ModeRFC4648
 static void DetectTransformFromBase64DecodeRegisterTests(void);
 #endif
-static void TransformFromBase64Decode(
-        DetectEngineThreadCtx *det_ctx, InspectionBuffer *buffer, void *options);
+#define B64D_ID_DATA_LEN 128
 
-void DetectTransformFromBase64DecodeRegister(void)
+static const char *id_format = "{"
+                               "\"flags\" : %d, "
+                               "\"bytes\" : { \"str\": \"%s\", \"val\": %d}, "
+                               "\"offset\" : { \"str\": \"%s\", \"val\": %d}, "
+                               "\"mode\" : %d"
+                               "}";
+
+static inline uint8_t *DetectTransformFromBase64IdData(
+        SCDetectTransformFromBase64Data *b64d, int *len)
 {
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].name = "from_base64";
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].desc = "convert the base64 decode of the buffer";
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].url = "/rules/transforms.html#from_base64";
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Setup = DetectTransformFromBase64DecodeSetup;
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Transform = TransformFromBase64Decode;
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Free = DetectTransformFromBase64DecodeFree;
-#ifdef UNITTESTS
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].RegisterTests =
-            DetectTransformFromBase64DecodeRegisterTests;
-#endif
-    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].flags |= SIGMATCH_OPTIONAL_OPT;
+    void *ptr = SCCalloc(1, B64D_ID_DATA_LEN);
+    if (ptr) {
+        int rc = snprintf(ptr, B64D_ID_DATA_LEN, id_format, b64d->flags,
```

Comment:
Do we actually need it (meaning the storage `id_data `) ? If it computed from fields in `SCDetectTransformFromBase64Data`, can we not compute it just in time ?

Reply:
Yes -- since there is an accessor function for fetching the ID.

What would be the benefit?

Reply:
Less memory usage and clearer code...

I think the first question for this and the 2 previous remarks is the encoding (printable string or binary)

Reply:
Could we just use the `SCDetectTransformFromBase64Data` stored in memory itself ?

Reply:
Yes -- that would work. Note that we can also do this following the merge of this work since the identity retrieval functions are in place.

Reply:
If I do this, than I should/will update the other transforms if the transform context contains the "identify" information, e.g., xor does (key) and pcrexform should as it's the compiled version of the regex.

Reply:
pcrexform looks fine for me as you do not know the layout of the pcre2 structure

Xor looks fine as well

Reply:
i'll update from_base64.

Reply:
> Could we just use the SCDetectTransformFromBase64Data stored in memory itself ?

That looks not the case as  `SCDetectTransformFromBase64Data` is not a flat structure because of the possible usage of variables for `offset` and `bytes` which are pointers to other buffers